cpp/src/rolling/rolling_detail.cuh

```diff
@@ -279,6 +282,92 @@ struct DeviceRollingCountAll {
+    : min_periods(_min_periods), ddof{_ddof}
+  {
+  }
+
+  template <typename OutputType, bool has_nulls>
+  bool __device__ operator()(column_device_view const& input,
+                             column_device_view const&,
+                             mutable_column_device_view& output,
+                             size_type start_index,
+                             size_type end_index,
+                             size_type current_index)
+  {
+    using DeviceInputType = device_storage_type_t<InputType>;
+
+    cudf::size_type count{0};  // valid counts in the window
+
+    // Count valid observations in window
+    if (has_nulls) {
+      count = thrust::count_if(thrust::seq,
+                               thrust::make_counting_iterator(start_index),
+                               thrust::make_counting_iterator(end_index),
+                               [&input](auto i) { return input.is_valid_nocheck(i); });
+    } else {
+      count = end_index - start_index;
+    }
+
+    // Variance/Std is non-negative, thus ddof should be strictly less than valid counts.
+    bool output_is_valid = (count >= min_periods) and not(count <= ddof);
+
+    if (output_is_valid) {
+      // Welford algorithm, a numerically stable, single pass algorithm
+      // See https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance
+      OutputType m, m2;
+      size_type running_count;
+
+      thrust::tie(running_count, m, m2) =
+        thrust::reduce(thrust::seq,
+                       thrust::make_counting_iterator(start_index),
+                       thrust::make_counting_iterator(end_index),
+                       thrust::make_tuple(size_type{0}, OutputType{0}, OutputType{0}),
+                       [&](auto acc, auto i) {
+                         if (has_nulls and input.is_null_nocheck(i)) { return acc; }
+
+                         OutputType m_acc, m2_acc, tmp1, tmp2;
+                         size_type r_count_acc;
+                         thrust::tie(r_count_acc, m_acc, m2_acc) = acc;
+                         OutputType x = static_cast<OutputType>(input.element<DeviceInputType>(i));
+
+                         r_count_acc++;
+                         tmp1 = x - m_acc;
+                         m_acc += tmp1 / r_count_acc;
+                         tmp2 = x - m_acc;
+                         m2_acc += tmp1 * tmp2;
+                         return thrust::make_tuple(r_count_acc, m_acc, m2_acc);
+                       });
+      if constexpr (is_fixed_point<InputType>()) {
+        // For fixed_point types, the previous computed value used unscaled rep-value,
+        // the final result should be multiplied by the square of decimal `scale`.
+        OutputType scaleby = exp10(static_cast<double>(input.type().scale()));
+        scaleby *= scaleby;
+        output.element<OutputType>(current_index) = m2 / (count - ddof) * scaleby;
+      } else {
+        output.element<OutputType>(current_index) = m2 / (count - ddof);
```

Comment:
From a spark perspective this may be one of the places where we have a small problem.

if `count == ddof` I belive that the output here will be `Inf`.

for `stddev_pop` this is probably fine. It is calculated using

```
    If(n === 0.0, Literal.create(null, DoubleType), sqrt(m2 / n))
```
where `n` is the same as `count` and we always set `min_periods` to 1 so output will not be valid if n == 0 and we get a null out like we want.

But for `stddev_samp` it is calculated using

```
    If(n === 0.0, Literal.create(null, DoubleType),
      If(n === 1.0, Double.NaN, sqrt(m2 / (n - 1.0))))
```

In this case if we get a single entry `n` will be 1 and we want the output to be `NaN`, but like I said before it looks like it will be `Inf`.  Is there any way we can special case this?

Reply:
I think we might be able to get away with replacing all `Inf` with `NaN`, but I think there is the possibility, although really rare, that m2 could overflow and become `Inf` and then we would end up with the wrong result.

Reply:
The calculation part here is guarded by:
```c++
    // Variance/Std is non-negative, thus ddof should be strictly less than valid counts.
    bool output_is_valid = (count >= min_periods) and not(count <= ddof);

    if (output_is_valid) {
    ...
    }
```
and the result is always null if `# valid observations <= ddof`.

In terms of m2 overflow, I don't have a good solution other than we use `__float128`, which is probably not supported natively on device. We just have to keep in mind that the result becomes unreliable if we encounter very large number that's on the order of `sqrt(np.finfo('float64').max) / N` (floating point number become inaccurate to that magnitude anyway).

Reply:
So the issue is that I need an output that matches as close as possible what Spark is doing.  I don't want to fix and m2 overflow, because Spark would also overflow for it.  I was just trying to figure out a cheap way to output the NaN instead of a null/Inf. And you are right it would be a null here. Now that I think about it I could always do a count aggregation with it and then do a copy if else in cases where I need to replace it.  So this should work. It just will not be as efficient as it coudl be for Spark.  But that is fine.

Reply:
Echoing the result of an offline discussion: in the case when `valid_count == ddof`, I'm more inclined to let the result follow standard c++ division (thus IEEE754) and leave that row as valid. The notion of "valid inputs should produce valid outputs" is quite persuasive to me. 

If I understand it correctly, this won't change the behavior on the Spark side - that you still need to do a count_valid and set these rows to be `NaN` accordingly. Let me know if I'm wrong.